track Half/ComplexHalf default dtype

bhosmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45263 use bitfield to shrink TensorImpl
* #45046 faster TensorOptions merging
* **#45043 track Half/ComplexHalf default dtype**
* #45026 pass TypeMeta by value

While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

Commit:
track Half/ComplexHalf default dtype

[ghstack-poisoned]

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]

ezyang commented on c10/core/DefaultDtype.cpp:
```diff
@@ -2,15 +2,21 @@
+  switch (default_dtype.toScalarType()) {
+    case ScalarType::Half:
+      default_complex_dtype = ScalarType::ComplexHalf;
+      break;
```
I mean, there's nothing wrong with adding this case, but it's not going to do anything cuz ComplexHalf really doesn't work at all haha.

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]

Commit:
Update on "track Half/ComplexHalf default dtype"


While tightening TypeMeta/ScalarType conversion, happened to notice that the default tracking wasn't observing the same relationship for Half/ComplexHalf as it was for Float and Double. This just adds that, in addition to updating the code. If the omission was intentional though just LMK and I'll drop it from the stack.

Differential Revision: [D23806097](https://our.internmc.facebook.com/intern/diff/D23806097)

[ghstack-poisoned]
